### ISSUE TITLE
systemd-oomd for Fedora CoreOS

### ISSUE THREAD

**Issue by dustymabe** [coreos/fedora-coreos-tracker#840]
**Describe the enhancement**

One Fedora 34 change was for [enabling `systemd-oomd` by default](https://fedoraproject.org/wiki/Changes/EnableSystemdOomd). My original understanding was that swap was required for this to work, but that no longer appears to be the case.

In an effort to keep differences from Fedora proper small we should consider enabling `systemd-oomd`.

**Comment by darkmuggle** [comment]
Per the meeting, @dustymabe and I will be investigating this.  This topic will be revisited at the next meeting.

**Comment by dustymabe** [comment]
Also @JaimeMagiera, who will be bringing it to the OKD working group.

@dghubble might have some input as well.

**Comment by dustymabe** [comment]
My action for this week:

`systemd-oomd` does not operate on systems that aren't cgroups v2. Though it does handle it gracefully by skipping startup with `ConditionControlGroupController=v2`: 

https://github.com/systemd/systemd-stable/blob/37c4cfde0ce613f0f00544d3f4e2e72bf93d9c76/units/systemd-oomd.service.in#L16

Also of interest: The defaults for Fedora are set in the `systemd-oomd-defaults` package. So we'd want to install that if we decide to enable this feature.

**Comment by travier** [comment]
So far, the OKD WG does not think this is useful for OKD as this might create issues: the kubelet would think a random process crashed and would restart the pod/application. @LorbusChris will reach out to G. Scrivano for more input.

See discussion in https://github.com/openshift/okd/discussions/663 and notes in https://hackmd.io/YJBn04R5TDi5Sm9XbOGwZA.

**Comment by LorbusChris** [comment]
@guiseppe do think systemd-oomd is useful in the case where most workloads are owned by the kubelet (i.e. OpenShift/OKD)?

**Comment by LorbusChris** [comment]
@giuseppe do think systemd-oomd is useful in the case where most workloads are owned by the kubelet (i.e. OpenShift/OKD)?

**Comment by giuseppe** [comment]
@LorbusChris I am not really sure as I've not played with it but I share the same concern as @travier: this could create confusion as both the kubelet and systemd-oomd on a OOM situation will try to kill processes and end up stepping on each other's toes.

**Comment by dustymabe** [comment]
Let's get kubernetes upstream (or individual downstreams) to [exclude the certain systemd units they want from systemd-oomd consideration](https://fedoraproject.org/wiki/Changes/EnableSystemdOomd#Can_we_exclude_certain_units_from_being_killed.3F).

**Comment by travier** [comment]
> Let's get kubernetes upstream (or individual downstreams) to [exclude the certain systemd units they want from systemd-oomd consideration](https://fedoraproject.org/wiki/Changes/EnableSystemdOomd#Can_we_exclude_certain_units_from_being_killed.3F).

I don't think that this is the behavior we want here. We probably don't want to exclude all pods from being killed as systemd-oomd would then target system daemons that we probably don't want to have killed either.

**Comment by jlebon** [comment]
Probably more appropriate to just have k8s/OKD disable systemd-oomd entirely instead?

**Comment by travier** [comment]
> Probably more appropriate to just have k8s/OKD disable systemd-oomd entirely instead?

That's what they are planning to do

**Comment by travier** [comment]
systemd-oomd is enabled in degraded mode (no swap) in Fedora Cloud:
```
systemd-oomd[510]: Swap is currently not detected; memory pressure usage will be degraded
```
and also enabled (with swap with default filesystem layout) in Fedora Server.

**Comment by dghubble** [comment]
I can't see how systemd-oomd would be useful on a Kubernetes node, where Kubelet already makes workload eviction decisions. I guess the question is, *is there a case for having kubelet manage workloads and systemd-oomd manage non-Kubernetes processes*. I think the two would clash or cause confusion depending on the timing of their actions.

**Comment by dustymabe** [comment]
We discussed this in the [community meeting today](https://lists.fedoraproject.org/archives/list/coreos@lists.fedoraproject.org/message/65OVFLTAS3BSVFKN2VNVFHF3SIKGSQNL/). A few relevant bits:

```
12:54:02      dustymabe | #info after talking to OKD, they plan to disable
                        | systemd-oomd. We stil need to talk to typhoon. In order to
                        | get more information from actual users we might end up
                        | enabling systemd-oomd or swap-on-zram+systemd-oomd in our
                        | next stream to get feedback on any potential issues.

```

Also:

```
12:55:18      dustymabe | #action dustymabe to add butane config to #840 to show
                        | how to enable systemd-oomd

```

**Comment by dustymabe** [comment]
> I can't see how systemd-oomd would be useful on a Kubernetes node, where Kubelet already makes workload eviction decisions. I guess the question is, _is there a case for having kubelet manage workloads and systemd-oomd manage non-Kubernetes processes_. I think the two would clash or cause confusion depending on the timing of their actions.

Thanks @dghubble for the Typhoon perspective. Since my suggestion in https://github.com/coreos/fedora-coreos-tracker/issues/840#issuecomment-857832415 wasn't a good one, an alternative may be to have FCOS default to disable `systemd-oomd` if you're running kubernetes.

Might be able to [achieve this](https://serverfault.com/questions/1016586/prevent-a-systemd-service-from-starting-if-another-systemd-service-is-running) with an `ExecCondition` like:

```
cat <<EOF > /etc/systemd/systemd-oomd.service.d/disable-on-kubernetes.conf
ExecCondition=/bin/bash -xc '/usr/sbin/systemctl is-enabled --quiet kubelet.service && exit 1 || exit 0'
EOF
```

Either way let's see how the exploratory testing that was proposed in the community meeting goes.

**Comment by cgwalters** [comment]
I think this is an upstream Kubernetes + systemd issue, not specific to FCOS (or OKD/OpenShift).  It probably makes the most sense to track it in upstream Kubernetes.  The conclusion may simply be to add a recommendation that systems installing kubelet also disable/mask `systemd-oomd.service`.  If you can install `kubelet.service` you can do that too. 

I don't see the need for us to carry a FCOS-specific thing to look for a hardcoded `kubelet.service` that doesn't exist by deafult.

**Comment by dghubble** [comment]
I'm happy to disable `systemd-oomd.service` in Typhoon to keep the OS from needing to have an ExecCondition for it. Seems ok to do now that every channel is on F34 and can be present, in case FCOS would like to enable it.

**Comment by cgwalters** [comment]
> I don't see the need for us to carry a FCOS-specific thing to look for a hardcoded kubelet.service that doesn't exist by deafult.

To elaborate though I do see the value in the *idea* of trying to be compatible in this space.  But if we're saying FCOS is independent of Kubernetes, then our special OS side hack isn't going to help someone who is e.g. doing something custom with podman or docker across upgrades.  

And so far no one has argued for making this one a "provisioning discontinuity" like cgroupsv2 (right?).

Another aspect to this is that having oomd enabled by default is unlikely to be instantly fatal - some systems may start failing in some cases on upgrades (admittedly in a way that may be hard to debug) but unlike cgroupsv2 it's not going to instantly break things that aren't ready for it.

So I think the strawman is:

 - Get the message out that oomd is coming and ensure higher level FCOS consumers have prepared if necessary
 - Enable by default (but maybe delay another Fedora release?)

**Comment by darkmuggle** [comment]
Fedora enables `systemd-oomd` for all variants[1] and the server uses the default profile.  During my deep dive on the question of OOM, my conclusion is that enabling swap by default would give us the bigger bang for the buck. The misconception about swap is that its a place to hold memory (or expand memory) but really swap is a place where inactive pages can be flushed to disk analogous to a file cached in memory. When a system is under memory pressure, it will try to free pages, and not having swap means that the pressure can result in OOM's quickly. `systemd-oomd` can terminate a badly behaving program, but without swap, we lose some burst-ability.  

I think we should enable it by default, but without some swap buffer, we're going to make entire pods subject to be killed on burst -- systemd-oomd kills the cgroup so people could have their well-behaved DB nuked because their $APP went a little crazy. 

[1] https://src.fedoraproject.org/rpms/fedora-release/blob/rawhide/f/90-default.preset
[2] https://src.fedoraproject.org/rpms/fedora-release/tree/rawhide

**Comment by dustymabe** [comment]
> @cgwalters 
> So I think the strawman is:> 
>     * Get the message out that oomd is coming and ensure higher level FCOS consumers have prepared if necessary
>     * Enable by default (but maybe delay another Fedora release?)

Yeah I think that's the idea. We'll initially do a round of exploratory testing amongst ourselves (i'll post a butane config in this issue). Subsequently maybe let it soak in `next` for a while before moving to `testing` and on to `stable`. 


> @darkmuggle 
> During my deep dive on the question of OOM, my conclusion is that enabling swap by default would give us the bigger bang for the buck.

Yeah I think we should start to revive the swap-on-zram discussion [since we said we'd revisit it again](https://github.com/coreos/fedora-coreos-tracker/issues/509#issuecomment-708614404). I'll probably start a new ticket for that.

**Comment by travier** [comment]
I took another look at what system-oomd does and I agree that should probably start the process of enabling it by default at least on new nodes with a transition period like we are doing for countme/cgroupsv2.

Looks like we need to seriously consider enabling swap on zram too.

**Comment by cgwalters** [comment]
> Yeah I think we should start to revive the swap-on-zram discussion since we said we'd revisit it again. I'll probably start a new ticket for that.

Right, swap-on-zram is (AIUI) much more *predictable* than traditional swap-on-block.  

(Because I am old, I remember seeing boxed software in Micro Center for (I think it was) a Windows 95 addon that was literally exactly compressed RAM...oh wow, right [the Internet never forgets](https://tidbits.com/2019/01/24/25-years-ago-in-tidbits-ram-doubler-debuts/).  I guess CPUs have gotten so much faster for some workloads it can make sense to burn a small portion of CPU to gain density)

**Comment by dustymabe** [comment]
> 
> ```
> 12:55:18      dustymabe | #action dustymabe to add butane config to #840 to show
>                         | how to enable systemd-oomd
> ```

A bit of an oversight on our part, but I just noticed it's already enabled in FCOS (brought in with systemd defaults in f34). So if you're running cgroups v2 (default for newly deployed nodes), you're already running it.

What we don't have is the `systemd-oomd-defaults` package that delivers some of the configuration.

The below config will enable swap-on-zram and install the `systemd-oomd-defaults` package.


```
variant: fcos
version: 1.3.0
storage:
  files:
    - path: /etc/systemd/zram-generator.conf
      mode: 0644
      contents:
        inline: |
          # This config file enables a /dev/zram0 device with the default settings
          [zram0]
systemd:
  units:
    - name: rpm-ostree-install.service
      enabled: true
      contents: |
        [Unit]
        Description=Layer rpms
        # We run after `systemd-machine-id-commit.service` to ensure that
        # `ConditionFirstBoot=true` services won't rerun on the next boot.
        After=systemd-machine-id-commit.service
        After=network-online.target
        ConditionPathExists=!/var/lib/rpm-ostree-install.stamp

        [Service]
        Type=oneshot
        RemainAfterExit=yes
        ExecStart=/usr/bin/rpm-ostree install --allow-inactive systemd-oomd-defaults
        ExecStart=/bin/touch /var/lib/rpm-ostree-install.stamp
        ExecStart=/bin/systemctl --no-block reboot

        [Install]
        WantedBy=multi-user.target
```

Please test! Either with or without zram would be nice.

**Comment by dghubble** [comment]
Maybe I've missed something, it doesn't seem enabled by default yet.

```
$ rpm-ostree status
State: idle
AutomaticUpdatesDriver: Zincati
  DriverState: active; periodically polling for updates (last checked Thu 2021-06-10 22:02:40 UTC)
Deployments:
● ostree://fedora:fedora/x86_64/coreos/testing
                   Version: 34.20210529.2.0 (2021-06-01T19:23:21Z)
                BaseCommit: d7ad41d882de1a9b5652d29ea69b0aedb83e5dec66cb4ce379ff651af14536ee
              GPGSignature: Valid signature by 8C5BA6990BDB26E19F2A1A801161AE6945719A39
           LayeredPackages: qemu-user-static

  ostree://fedora:fedora/x86_64/coreos/testing
                   Version: 34.20210529.2.0 (2021-06-01T19:23:21Z)
                    Commit: d7ad41d882de1a9b5652d29ea69b0aedb83e5dec66cb4ce379ff651af14536ee
              GPGSignature: Valid signature by 8C5BA6990BDB26E19F2A1A801161AE6945719A39
$ systemctl status systemd-oomd
○ systemd-oomd.service - Userspace Out-Of-Memory (OOM) Killer
     Loaded: loaded (/usr/lib/systemd/system/systemd-oomd.service; disabled; vendor preset: disabled)
     Active: inactive (dead)
       Docs: man:systemd-oomd.service(8)
$ grep cgroup /proc/filesystems
nodev   cgroup
nodev   cgroup2
```

**Comment by dustymabe** [comment]
> Maybe I've missed something, it doesn't seem enabled by default yet.


You're right. I think the system I checked on was one I was using to explore `systemd-oomd` a few weeks back.

**Comment by cgwalters** [comment]
> You're right. I think the system I checked on was one I was using to explore systemd-oomd a few weeks back.

And that knowledge went...out of your memory eh?

**Comment by dustymabe** [comment]
Totally evicted!

**Comment by dustymabe** [comment]
> > Maybe I've missed something, it doesn't seem enabled by default yet.
> 
> You're right. I think the system I checked on was one I was using to explore `systemd-oomd` a few weeks back.

I think something else that added to my confusion is that the butane config I added in https://github.com/coreos/fedora-coreos-tracker/issues/840#issuecomment-859094939 did work even though I didn't explicitly enable `systemd-oomd`. Just adding the `systemd-oomd-defaults` package makes it start for some reason, though it's not clear to me how.

**Comment by dustymabe** [comment]
> Just adding the `systemd-oomd-defaults` package makes it start for some reason, though it's not clear to me how.

@miabbott pointed out to me that adding configuration (like `ManagedOOMMemoryPressure=kill`) to any systemd unit will activate `systemd-oomd`. We'll need to mask it in cases where we don't want it to be run.

**Comment by travier** [comment]
> > Just adding the `systemd-oomd-defaults` package makes it start for some reason, though it's not clear to me how.
> 
> @miabbott pointed out to me that adding configuration (like `ManagedOOMMemoryPressure=kill`) to any systemd unit will activate `systemd-oomd`. We'll need to mask it in cases where we don't want it to be run.

OK, this is a point for enabling (or not masking) by default as potentially expected functionality will break if we mask the unit.

**Comment by dustymabe** [comment]
> OK, this is a point for enabling (or not masking) by default as potentially expected functionality will break if we mask the unit.

yeah, though we should find and fix those bugs in systemd/oomd if they exist. User's should be able to mask it without side effects.

**Comment by dustymabe** [comment]
We discussed this in [the community meeting today](https://lists.fedoraproject.org/archives/list/coreos@lists.fedoraproject.org/message/4IRREF6EHWWRXX5C7A56KX7ZEX74A7BR/).

@jdoss is working on doing some testing and we'll hear back from him next week.

We did make a small decision:

```
  * AGREED: since oomd works better with swap, let's tie the swaponzram
    proposal and the oomd proposals together. If we do one, we do the
    other.  (dustymabe, 16:50:58)
```

but we also decided to take a step back and discuss single node versus kubernetes defaults briefly first: https://github.com/coreos/fedora-coreos-tracker/issues/880

**Comment by andygeorge** [comment]
> @jdoss is working on doing some testing

(below hosts are both on `testing-devel/builds/34.20210620.20.0`)

We recently ran into some memory leaks in our core app that caused the app (and eventually the fcos host) to go unresponsive. `oom-killer` did do some work, but did not kill the container causing the memory leak, which prevented our app from recovering automatically.

**no swap-on-zram or systemd-oomd-defaults**:
![image](https://user-images.githubusercontent.com/10601857/123304729-e5533680-d4e4-11eb-9e4b-7f3965282d25.png)

We just tested this against another host with swap-on-zram enabled + systemd-oomd-defaults installed (using @dustymabe's [config](https://github.com/coreos/fedora-coreos-tracker/issues/840#issuecomment-859094939); thank you for that unit!): swap-on-zram worked great and the systemd-oomd (w/defaults) killed our offending container as hoped, without any adverse affects on our app.

**swap-on-zram enabled + systemd-oomd-defaults installed**:
![image](https://user-images.githubusercontent.com/10601857/123304742-eab08100-d4e4-11eb-8766-29a2d53f6f21.png)


...with a very small/handy entry that we'll be able to easily use for logging/alerting to catch the next memory leak (maybe this will be our last! ha ha):
```
[core@community (community.testhost.com) ~]$ journalctl -u systemd-oomd -f
-- Journal begins at Thu 2021-06-24 14:06:29 UTC. --
Jun 24 14:18:27 community.testhost.com systemd[1]: Starting Userspace Out-Of-Memory (OOM) Killer...
Jun 24 14:18:28 community.testhost.com systemd[1]: Started Userspace Out-Of-Memory (OOM) Killer.
Jun 24 16:35:32 community.testhost.com systemd-oomd[3455]: Killed /machine.slice/machine-libpod_pod_5cd150410f0f8fd2825bd8232e212ec151f1aaa9e1899762efa5012ebdc522f4.slice/libpod-a20b33b96087e0ce7637845abde6efedee3bd6ac233f4ba98f1eadc4512a51f5.scope/container due to swap used (1854169088) / total (2057302016) being more than 90.00%
```

From our perspective, this is definitely a nicer experience with those two enabled.

**Comment by cmurf** [comment]
Maybe it's time to revist this now?

**Comment by jdoss** [comment]
I agree. We should enable it by default.

**Comment by miabbott** [comment]
Related: https://lists.fedoraproject.org/archives/list/devel@lists.fedoraproject.org/thread/CRREDQUPPJYWVRMA4DOKYU2KZZLKC4D5/

**Comment by kannon92** [comment]
I wanted to play around with installing systemd-oomd on FCOS. Has anyone found any useful documentation on what is needed for that?

I've actually been making a bit of progress of investigating how one could use systemd-oomd, swap and kubernetes on fedora. But not really sure how one installs this on FCOS.

**Comment by dustymabe** [comment]
Note systemd-oomd enablement slipped into the latest `testing` stream `43.20260413.2.1` (also latest `next`) because the `systemd-oomd.socket` [got enabled in the presets](https://src.fedoraproject.org/rpms/fedora-release/c/ce9c951bf62beea766f405774b4948687f153881?branch=f43).

At this point no one has complained about it being enabled in `testing` or `next`, but it also hasn't landed in `stable` yet.

The open question:

- Do we take this as a sign that systemd-oomd doesn't break anything (even without swap) and allow it to continue?
- OR do we add a commit to the `stable` when we promote to ensure it doesn't reach `stable` users.


<details>

Tested this using the following script with coreos-builds-bisect.py:

```
[dustymabe@media fcos-buildfetch]$ cat test-gcp.sh 
#!/bin/bash

set -eu

cosa() {
   podman run --rm -i --security-opt label=disable --privileged      \
       --userns keep-id:uid=1000,gid=1000                            \
       -v ${PWD}:/srv/ --device /dev/kvm --device /dev/fuse          \
       --tmpfs /tmp -v /var/tmp:/var/tmp --name cosa                 \
       quay.io/coreos-assembler/coreos-assembler:latest "$@"
}

build=$1
arch='x86_64'
cosa buildfetch --arch=$arch --build=$build

output=$(cosa kola spawn --build=$build -p=gcp --gcp-json-key=./gcp-key \
        --gcp-project=fedora-coreos-devel --ssh-command "systemctl show --property=ActiveState systemd-oomd.service")

echo "${output}"

if grep --quiet 'ActiveState=active' <<< "${output}"; then
    echo "${build} is BAD"
    exit 1
fi

echo "${build} is GOOD"
```


```
$ ./coreos-builds-bisect.py --good=43.20260227.20.1 --bad=43.20260419.20.1 --testscript=./test-gcp.sh 
Using data file at: ./coreos-builds-bisect-data.json
Targetting architecture: x86_64                       
Executing test for new build: 43.20260327.20.1. ~5 steps left
...
...
43.20260411.20.dev0 is GOOD           
43.20260411.20.dev0 passed                                                                                                                             
BISECT TEST RESULTS:
Last known good build: 43.20260411.20.dev0
First known bad build: 43.20260413.20.1
```

$ cosa diff --rpms --from=43.20260411.20.dev0 --to=43.20260413.20.1                                                   
...
ostree diff commit from: 43.20260411.20.dev0 (6ca25e40171a272da66aee26bd81a77587b1af54038332da2bb93e80fe894e4f)
ostree diff commit to:   43.20260413.20.1 (890fcd8aad401abbb2b953b4f82bc2907cf714b82b2272bab91a69ca2d867d0a)                                           
Upgraded:                                                                                                                                              
  fedora-release-common 43-26 -> 43-27                                                                                                                 
  fedora-release-coreos 43-26 -> 43-27                                                                                                                                                                                                                                                                        
  fedora-release-identity-coreos 43-26 -> 43-27
  glibc 2.42-10.fc43 -> 2.42-11.fc43  
  glibc-common 2.42-10.fc43 -> 2.42-11.fc43                                                                                                            
  glibc-gconv-extra 2.42-10.fc43 -> 2.42-11.fc43
  glibc-minimal-langpack 2.42-10.fc43 -> 2.42-11.fc43
  libcap 2.76-3.fc43 -> 2.76-4.fc43
```

</details>

**Comment by travier** [comment]
> Is there still plans to include this by default?

The `systemd-oomd` service is included in Fedora CoreOS already, only disabled. You can enable it via Ignition or manually.

`zram-generator` (for https://github.com/coreos/fedora-coreos-tracker/issues/859) is also included but not enabled by default: https://docs.fedoraproject.org/en-US/fedora-coreos/sysconfig-configure-swaponzram/

**Comment by kannon92** [comment]
https://github.com/coreos/fedora-coreos-tracker/issues/840#issuecomment-4297658918

The main concern I have is that projects like Kubernetes haven't really begun to think through how to support systemd-oomd.

I did some minor playing around with it and the UX is quite poor. (This was with Swap though). 

I don't think this needs to block your decision on enabling or disabling it though but just wanted to mention this.

**Comment by prestist** [comment]
We discussed this in the meeting today and came to the agreement that  **we will undo the change and make sure the next releases of CoreOS don't ship with systemd-oomd enabled** effectively we will protect stable, and block this change from promoting its self to stable. 

[Meeting Log ](https://meetbot-raw.fedoraproject.org//meeting-1_matrix_fedoraproject-org/2026-04-22/fedora-coreos-meeting.2026-04-22-15.37.log.txt)

**Comment by dustymabe** [comment]
Opened https://github.com/coreos/fedora-coreos-config/pull/4122
